pkg/target/target_test.go

```diff
@@ -693,13 +695,74 @@ func TestMatcher_Match(t *testing.T) {
 	}
 }
 
+func TestNamespaceCache(t *testing.T) {
+	ns1 := makeNamespace("my-ns1", map[string]string{"ns1": "label"})
+	ns2 := makeNamespace("my-ns2", map[string]string{"ns2": "label"})
+	tests := []struct {
+		name    string
+		addNs   map[string]interface{}
+		wantNs  map[string]*corev1.Namespace
+		wantErr bool
+	}{
+		{
+			name:    "retrieving a namespace from empty cache returns nil",
+			addNs:   map[string]interface{}{},
+			wantNs:  map[string]*corev1.Namespace{"my-ns2": nil},
+			wantErr: false,
+		},
+		{
+			name:    "retrieving a namespace that doesnt exist returns nil",
+			addNs:   map[string]interface{}{"my-ns1": ns1},
+			wantNs:  map[string]*corev1.Namespace{"my-ns2": nil},
+			wantErr: false,
+		},
+		{
+			name:    "retrieving an added namespace returns the namespace",
+			addNs:   map[string]interface{}{"my-ns1": ns1, "my-ns2": ns2},
+			wantNs:  map[string]*corev1.Namespace{"my-ns1": ns1, "my-ns2": ns2},
+			wantErr: false,
+		},
+		{
+			name:    "adding a non-namespace type returns error",
+			addNs:   map[string]interface{}{"my-ns1": fooConstraint(), "my-ns2": ns2},
+			wantNs:  map[string]*corev1.Namespace{"my-ns2": ns2},
+			wantErr: true,
+		},
+	}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			cache := newNsCache(nil)
+
+			for key, ns := range tt.addNs {
+				err := cache.Add(key, ns)
+				if !tt.wantErr && err != nil {
+					t.Errorf("unwanted error adding namespace")
+				}
+			}
+
+			for key, want := range tt.wantNs {
+				got, err := cache.Get(key)
+				if !tt.wantErr && err != nil {
+					t.Errorf("unwanted error getting namespace")
+				}
+				if want == nil && got == nil {
+					continue
+				}
+				if diff := cmp.Diff(got, want); diff != "" {
+					t.Errorf("Get() got = %v, want %v", got, want)
```

Comment:
Use `t.Error(diff)` or the longer `t.Errorf("+got -want:\n%s", diff)` so we get the pretty-printed diff.